File: VB/WindowsApplication41/RatingControl.vb
Imports System
Imports System.Drawing
Imports System.Drawing.Drawing2D
Imports System.ComponentModel
Imports System.Windows.Forms
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraPrinting
Imports DevExpress.Utils.Drawing
Imports DevExpress.Utils

<ToolboxItem(True)>
Public Class RatingControl
    Inherits BaseEdit

    Shared Sub New()
        RepositoryItemRatingControl.RegisterMyRatingControl()
    End Sub

    Public Sub New()
        TabStop = False
        fEditValue = 0
        fOldEditValue = fEditValue
    End Sub

    <Browsable(False)>
    Public Overrides ReadOnly Property EditorTypeName As String
        Get
            Return "RatingControl"
        End Get
    End Property

    <Description("Gets an object containing properties, methods and events specific to rating controls."), Category(CategoryName.Properties), DesignerSerializationVisibility(DesignerSerializationVisibility.Content)>
    Public Overloads ReadOnly Property Properties As RepositoryItemRatingControl
        Get
            Return TryCast(MyBase.Properties, RepositoryItemRatingControl)
        End Get
    End Property

    <Browsable(False), DefaultValue(0)>
    Public Overrides Property EditValue As Object
        Get
            Return MyBase.EditValue
        End Get

        Set(ByVal value As Object)
            MyBase.EditValue = ConvertCheckValue(value)
        End Set
    End Property

    <Category(CategoryName.Appearance), Description("Gets or sets position."), RefreshProperties(RefreshProperties.All), DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden), DefaultValue(0), Bindable(ControlConstants.NonObjectBindable)>
    Public Overridable Property Position As Integer
        Get
            Return Properties.ConvertValue(EditValue)
        End Get

        Set(ByVal value As Integer)
            EditValue = Properties.CheckValue(value)
        End Set
    End Property

    Public Overridable Property Title As String
        Get
            Return Properties.Title
        End Get

        Set(ByVal value As String)
            EditValue = value
        End Set
    End Property

    Protected Overridable Function ConvertCheckValue(ByVal val As Object) As Object
        If val Is Nothing OrElse TypeOf val Is DBNull Then Return val
        Dim converted As Integer = Properties.ConvertValue(val)
        Try
            If converted = Convert.ToInt32(val) Then Return val
        Catch
        End Try

        Return converted
    End Function

    Protected Friend Overridable Sub OnMinMaxChanged()
        If IsLoading Then Return
        Position = Properties.CheckValue(Position)
    End Sub

    Protected Friend Overridable Sub UpdatePercent()
        Dim vi As RatingControlViewInfo = TryCast(ViewInfo, RatingControlViewInfo)
        If vi Is Nothing Then Return
        vi.UpdateProgressInfo(vi.ProgressInfo)
    End Sub

    Public Custom Event PositionChanged As EventHandler
        AddHandler(ByVal value As EventHandler)
            AddHandler Properties.PositionChanged, value
        End AddHandler

        RemoveHandler(ByVal value As EventHandler)
            RemoveHandler Properties.PositionChanged, value
        End RemoveHandler

        RaiseEvent(ByVal sender As Object, ByVal e As EventArgs)
        End RaiseEvent
    End Event

    Protected Overrides Sub OnMouseMove(ByVal e As MouseEventArgs)
        Properties.HotTrackIndex = GetValueByPoint(e.Location)
        UpdateValue()
        MyBase.OnMouseMove(e)
    End Sub

    Private Function GetValueByPoint(ByVal location As Point) As Integer
        Dim steps As Double = CDbl(Properties.Maximum) / CDbl(Width)
        Dim value As Double = location.X * steps
        Return CInt(Math.Ceiling(value))
    End Function

    Protected Overrides Sub OnMouseClick(ByVal e As MouseEventArgs)
        Position = GetValueByPoint(e.Location)
        UpdateValue()
        MyBase.OnMouseClick(e)
    End Sub

    Protected Overrides Sub OnMouseLeave(ByVal e As EventArgs)
        'TODO:
        Properties.HotTrackIndex = 0
        MyBase.OnMouseLeave(e)
    End Sub

    Protected Friend Overridable Sub UpdateValue()
        Dim vi As RatingControlViewInfo = TryCast(ViewInfo, RatingControlViewInfo)
        If vi Is Nothing Then Return
        vi.UpdateProgressInfo(vi.ProgressInfo)
    End Sub
End Class

File: CS/WindowsApplication41/RatingControl.cs
using System;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Collections;
using System.ComponentModel;
using System.Windows.Forms;
using DevExpress.LookAndFeel;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.ViewInfo;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraPrinting;
using DevExpress.XtraEditors.Registrator;
using DevExpress.Utils.Drawing;
using DevExpress.Utils;

[ToolboxItem(true)]
public class RatingControl : BaseEdit {

    static RatingControl() {
        RepositoryItemRatingControl.RegisterMyRatingControl();
    }
    public RatingControl() {
        this.TabStop = false;
        this.fOldEditValue = this.fEditValue = 0;
    }
    [Browsable(false)]
    public override string EditorTypeName { get { return "RatingControl"; } }
    [Description("Gets an object containing properties, methods and events specific to rating controls."), Category(CategoryName.Properties), DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
    public new RepositoryItemRatingControl Properties { get { return base.Properties as RepositoryItemRatingControl; } }
    [Browsable(false), DefaultValue(0)]
    public override object EditValue {
        get { return base.EditValue; }
        set { base.EditValue = ConvertCheckValue(value); }
    }
    [Category(CategoryName.Appearance), Description("Gets or sets position."), RefreshProperties(RefreshProperties.All), DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden), DefaultValue(0),
    Bindable(ControlConstants.NonObjectBindable)]
    public virtual int Position {
        get { return Properties.ConvertValue(EditValue); }
        set { EditValue = Properties.CheckValue(value); }
    }
    public virtual string Title {
        get { return Properties.Title; }
        set { EditValue = value; }
    }
    protected virtual object ConvertCheckValue(object val) {
        if(val == null || val is DBNull) return val;
        int converted = Properties.ConvertValue(val);
        try {
            if(converted == Convert.ToInt32(val))
                return val;
        } catch { }
        return converted;
    }
    protected internal virtual void OnMinMaxChanged() {
        if(IsLoading) return;
        Position = Properties.CheckValue(Position);
    }
    protected internal virtual void UpdatePercent() {
        RatingControlViewInfo vi = ViewInfo as RatingControlViewInfo;
        if(vi == null) return;
        vi.UpdateProgressInfo(vi.ProgressInfo);
    }
    public event EventHandler PositionChanged {
        add { Properties.PositionChanged += value; }
        remove { Properties.PositionChanged -= value; }
    }
    protected override void OnMouseMove(MouseEventArgs e) {
        Properties.HotTrackIndex = GetValueByPoint(e.Location);
        UpdateValue();
        base.OnMouseMove(e);
    }
    private int GetValueByPoint(Point location) {
        double steps = (double)Properties.Maximum / (double)Width;
        double value = location.X * steps;
        return (int)Math.Ceiling(value);
    }
    protected override void OnMouseClick(MouseEventArgs e) {
        Position = GetValueByPoint(e.Location);
        UpdateValue();
        base.OnMouseClick(e);
    }
    protected override void OnMouseLeave(EventArgs e) {
        //TODO:
        Properties.HotTrackIndex = 0;
        base.OnMouseLeave(e);
    }
    protected internal virtual void UpdateValue() {
        RatingControlViewInfo vi = ViewInfo as RatingControlViewInfo;
        if(vi == null) return;
        vi.UpdateProgressInfo(vi.ProgressInfo);
    }
}
